File: C++/Parameter_BinningHV1/Grab_BinningHV1.cpp
/*
	BinningHV1: カメラの水平ビニング、垂直ビニングに1を設定
	BinningHV1: Set Camera Binning Horizontal / Binning Vertical to 1
*/

#define ENABLED_ST_GUI

#include <StApi_TL.h>
#ifdef ENABLED_ST_GUI
#include <StApi_GUI.h>
#include <iomanip>	//std::setprecision
#endif

using namespace StApi;
using namespace std;
const uint64_t nCountOfImagesToGrab = 100;

int main(int /* argc */, char ** /* argv */)
{
	try
	{
		CStApiAutoInit objStApiAutoInit;
		CIStSystemPtr pIStSystem(CreateIStSystem(StSystemVendor_Sentech));
		CIStDevicePtr pIStDevice(pIStSystem->CreateFirstIStDevice());
		cout << "Device=" << pIStDevice->GetIStDeviceInfo()->GetDisplayName() << endl;

#ifdef ENABLED_ST_GUI
		CIStImageDisplayWndPtr pIStImageDisplayWnd(CreateIStWnd(StWindowType_ImageDisplay));
#endif
		CIStDataStreamPtr pIStDataStream(pIStDevice->CreateIStDataStream(0));

		// ==============================================================================================================
		// 水平、垂直Binningに1を設定するデモ
		// Demostration of setting Binning Horizontal / Binning Veritcal to 1

		// パラメータにアクセスするためのノードマップポインタを生成
		// Create NodeMap pointer for accessing parameters
		GenApi::CNodeMapPtr pNodeMapCameraParam(pIStDevice->GetRemoteIStPort()->GetINodeMap());

		// BinningHorizontalのノードを取得
		// Get Node for Binning Horizontal
		GenApi::CNodePtr pNodeBinningH = pNodeMapCameraParam->GetNode("BinningHorizontal");
		// 値を設定するためにNodeをCIntegerPtrに変換
		// Convert Node to CIntegerPtr for setting value
		GenApi::CIntegerPtr pIntBinningH(pNodeBinningH);
		// BinningHorizontalに1を設定
		// Set BinningHorizontal to 1
		pIntBinningH->SetValue(1);

		// BinningVerticalのノードを取得
		// Get Node for Binning Vertical
		GenApi::CNodePtr pNodeBinningV = pNodeMapCameraParam->GetNode("BinningVertical");
		// 値を設定するためにNodeをCIntegerPtrに変換
		// Convert Node to CIntegerPtr for setting value
		GenApi::CIntegerPtr pIntBinningV(pNodeBinningV);
		// BinningVerticalに1を設定
		// Set BinningVertical to 1
		pIntBinningV->SetValue(1);

		// Binningに1を設定するためにHeight、Widthを手動で最大値に設定する必要がある。
		// For setting Binning to 1, both height and width need to be set to max manually.

		// Widthノードを取得
		// Get Node for Width
		GenApi::CNodePtr pNodeWidth = pNodeMapCameraParam->GetNode("Width");
		// 値を設定するためにNodeをCIntegerPtrに変換
		// Convert Node to CIntegerPtr for setting value
		GenApi::CIntegerPtr pIntWidthl(pNodeWidth);
		// Widthに可能な限り最大値を設定
		// Set Width to Max as it can
		pIntWidthl->SetValue(pIntWidthl->GetMax());

		// Heightノードを取得
		// Get Node for Height
		GenApi::CNodePtr pNodeHeight = pNodeMapCameraParam->GetNode("Height");
		// 値を設定するためにNodeをCIntegerPtrに変換
		// Convert Node to CIntegerPtr for setting value
		GenApi::CIntegerPtr pIntHeight(pNodeHeight);
		// Heightに可能な限り最大値を設定
		// Set Height to Max as it can
		pIntHeight->SetValue(pIntHeight->GetMax());

		// ==============================================================================================================

		pIStDataStream->StartAcquisition(nCountOfImagesToGrab);
		pIStDevice->AcquisitionStart();
		while (pIStDataStream->IsGrabbing())
		{
			CIStStreamBufferPtr pIStStreamBuffer(pIStDataStream->RetrieveBuffer(5000));
			if (pIStStreamBuffer->GetIStStreamBufferInfo()->IsImagePresent())
			{
				IStImage *pIStImage = pIStStreamBuffer->GetIStImage();

#ifdef ENABLED_ST_GUI
				stringstream ss;
				ss << pIStDevice->GetIStDeviceInfo()->GetDisplayName();
				ss << "  ";
				ss << pIStImage->GetImageWidth() << " x " << pIStImage->GetImageHeight();
				ss << "  ";
				ss << fixed << std::setprecision(2) << pIStDataStream->GetCurrentFPS();
				ss << "[fps]";
                GenICam::gcstring strText(ss.str().c_str());
				pIStImageDisplayWnd->SetUserStatusBarText(strText);

				if (!pIStImageDisplayWnd->IsVisible())
				{
					pIStImageDisplayWnd->SetPosition(0, 0, pIStImage->GetImageWidth(), pIStImage->GetImageHeight());
					pIStImageDisplayWnd->Show(NULL, StWindowMode_ModalessOnNewThread);
				}

				pIStImageDisplayWnd->RegisterIStImage(pIStImage);
#else
				cout << "BlockId=" << pIStStreamBuffer->GetIStStreamBufferInfo()->GetFrameID()
					<< " Size:" << pIStImage->GetImageWidth() << " x " << pIStImage->GetImageHeight()
					<< " First byte =" << (uint32_t)*(uint8_t*)pIStImage->GetImageBuffer() << endl;
#endif
			}
			else
			{
				cout << "Image data does not exist" << endl;
			}
		}
		pIStDevice->AcquisitionStop();
		pIStDataStream->StopAcquisition();
	}
	catch (const GenICam::GenericException &e)
	{
		cerr << endl << "An exception occurred." << endl << e.GetDescription() << endl;
	}

	cout << endl << "Press Enter to exit." << endl;
	while (cin.get() != '\n');

	return(0);
}

File: C#/Parameter_BinningHV1CS/Grab_BinningHV1.cs
/*
	BinningHV1: カメラの水平ビニング、垂直ビニングに1を設定
	BinningHV1: Set Camera Binning Horizontal / Binning Vertical to 1
*/

#define ENABLED_ST_GUI

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Sentech.GenApiDotNET;
using Sentech.StApiDotNET;

namespace Parameter_BinningHV1
{
    class Grab_BinningHV1
    {
        const int nCountOfImagesToGrab = 100;
        static void Main(string[] args)
        {
            try
            {
                using (CStApiAutoInit api = new CStApiAutoInit())

                using (CStSystem system = new CStSystem(eStSystemVendor.Sentech))

                using (CStDevice device = system.CreateFirstStDevice())

                using (CStImageDisplayWnd wnd = new CStImageDisplayWnd())

                using (CStDataStream dataStream = device.CreateStDataStream(0))
                {
                    Console.WriteLine("Device=" + device.GetIStDeviceInfo().DisplayName);

                    // ===========================================================================================================
                    // 水平、垂直Binningに1を設定するデモ
                    // Demostration of setting Binning Horizontal / Binning Veritcal to 1

                    // パラメータにアクセスするためのノードマップポインタを生成
                    // Create NodeMap pointer for accessing parameters
                    INodeMap nodeMap = device.GetRemoteIStPort().GetINodeMap();

                    // BinningHorizontalのノードを取得
                    // Acquire Binning Horizontal node
                    IInteger intBinningHorizontal = nodeMap.GetNode<IInteger>("BinningHorizontal");
                    // BinningHorizontalに1を設定
                    // Set BinningHorizontal to 1
                    intBinningHorizontal.Value = 1;

                    // BinningVerticalのノードを取得
                    // Acquire Binning Vertical node
                    IInteger intBinningVertical = nodeMap.GetNode<IInteger>("BinningVertical");
                    // BinningVerticalに1を設定
                    // Set BinningVertical to 1
                    intBinningVertical.Value = 1;

                    // Binningに1を設定するためにHeight、Widthを手動で最大値に設定する必要がある。
                    // For setting Binning to 1, both height and width need to be set to max manually.

                    // Widthノードを取得
                    // Acquire Width node
                    IInteger intWidth = nodeMap.GetNode<IInteger>("Width");
                    // Widthに可能な限り最大値を設定
                    // Set Width to Max as it can
                    intWidth.Value = intWidth.Maximum;

                    // Heightノードを取得
                    // Acquire Height node
                    IInteger intHeight = nodeMap.GetNode<IInteger>("Height");
                    // Heightに可能な限り最大値を設定
                    // Set Height to Max as it can
                    intHeight.Value = intHeight.Maximum;
                    // ===========================================================================================================

                    dataStream.StartAcquisition(nCountOfImagesToGrab);

                    device.AcquisitionStart();

                    while (dataStream.IsGrabbing)
                    {
                        using (CStStreamBuffer streamBuffer = dataStream.RetrieveBuffer(5000))
                        {
                            if (streamBuffer.GetIStStreamBufferInfo().IsImagePresent)
                            {
                                IStImage stImage = streamBuffer.GetIStImage();
                                string strText = device.GetIStDeviceInfo().DisplayName + " ";
                                strText += stImage.ImageWidth + " x " + stImage.ImageHeight + " ";
                                strText += string.Format("{0:F2}[fps]", dataStream.CurrentFPS);
                                wnd.SetUserStatusBarText(strText);

                                if (!wnd.IsVisible)
                                {
                                    wnd.SetPosition(0, 0, (int)stImage.ImageWidth, (int)stImage.ImageHeight);

                                    wnd.Show(eStWindowMode.ModalessOnNewThread);
                                }

                                wnd.RegisterIStImage(stImage);
                            }
                            else
                            {
                                Console.WriteLine("Image data does not exist.");
                            }
                        }
                    }

                    device.AcquisitionStop();

                    dataStream.StopAcquisition();
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("An exception occurred. \r\n" + e.Message);
            }
            finally
            {
                Console.WriteLine("\r\nPress Enter to exit.");
                Console.ReadLine();
            }
        }
    }
}
